torch/csrc/jit/codegen/cuda/kernel.cpp

```diff
@@ -25,58 +26,135 @@ static int ceilDiv(const int a, const int b) {
   return (a + b - 1) / b;
 }
 
+// IO data structure for kernel code;
+static auto typeinfo = R"(
+template<typename T, int N>
+struct Tensor {
+  T& operator[](int ind) {
+    return data[ind];
+  };
+
+  T* data;
+  int size[N];
+  int stride[N];
+};
+)";
+
+// include IO data structure for host code
+struct KernelArgumentHolder {
+  std::vector<void*> arguments;
+  std::vector<char> buffer;
+  void* buffer_ptr;
+
+  KernelArgumentHolder(size_t n_inputs, size_t n_dimension_per_tensor) {
+    arguments.reserve(n_inputs);
+    // We are being generous here on the allocated buffer;
+    buffer.resize(
+        n_inputs * (sizeof(void*) + 2 * sizeof(int) * n_dimension_per_tensor));
+    buffer_ptr = buffer.data();
+  }
+
+  void** args() {
+    return arguments.data();
+  }
+
+  // this should comply to the storage of Tensor object defined in typeinfo;
+  void push_tensor(
+      const at::Tensor& val,
+      c10::optional<at::IntArrayRef> broadcasted_size = c10::nullopt) {
+    arguments.push_back(buffer_ptr);
+
+    // passing address, type doesn't really matter here;
+    auto data_ptr = static_cast<void**>(buffer_ptr);
+    *data_ptr = val.data_ptr();
+    buffer_ptr += sizeof(char*);
+
+    if (broadcasted_size) {
+      auto b_dim = broadcasted_size->size();
+      auto o_dim = val.dim();
+      TORCH_CHECK(b_dim >= o_dim);
+      int* sizes = reinterpret_cast<int*>(buffer_ptr);
+      int* strides = &sizes[b_dim];
+      for (int i = 0; i < b_dim; i++) {
+        sizes[i] = broadcasted_size->at(i);
+        int index = i + o_dim - b_dim;
+        if (index < 0) {
```

Comment:
Is this assuming broadcasted dims are at the beginning of the tensor? I might just be misreading this.

Reply:
Yes it is. Numpy & PyTorch follow the same implicit broadcasting rule.

Reply:
So broadcast on tensors with mismatched dims (no 1-size place holder dims) can only be automated expanded at the beginning dims of the tensor?
i.e.
(4, 2, 4) can't be expanded with (4, 4)? but (4, 2, 4) and (4, 1, 4) can be?